### PR TITLE
mgr/dashboard: Add "nvmeof top connection" cmd

### PR DESCRIPTION
**Pull request by VallariAg**
Add a new command "nvmeof top connection" to
show host connection iostat.

Fixes: https://tracker.ceph.com/issues/78248

## Calculation Logic

Here's what I get from `connection get_io_statistics` cli:
```
[root@dev-vallari-cluster-node1 ~]# docker run -it quay.io/ceph/nvmeof-cli:1.9 --server-address 10.243.64.86 connection get_io_statistics -n 'nqn.2016-06.io.spdk:cnode1.mygroup1' -t 'nqn.2014-08.org.nvmexpress:uuid:6def1fb3-a3a2-447b-9938-59856fc4805f'
IO statistics for host nqn.2014-08.org.nvmexpress:uuid:6def1fb3-a3a2-447b-9938-59856fc4805f on nqn.2016-06.io.spdk:cnode1.mygroup1:

╒══════════╤══════════╤═════════════╤══════════════════╤══════════════════╤══════════════════╤══════════════════╕
│ Bucket   │ Bucket   │   IOs Count │ BDEV µSec        │ Net µSec         │ QOS µSec         │ Total µSec       │
│ Size     │ Type     │             │ (Min,Max,Mean)   │ (Min,Max,Mean)   │ (Min,Max,Mean)   │ (Min,Max,Mean)   │
╞══════════╪══════════╪═════════════╪══════════════════╪══════════════════╪══════════════════╪══════════════════╡
│ 2KB      │ Write    │           4 │ 303, 11599, 4774 │ 0, 0, 0          │ 0, 3, 2          │ 306, 11600, 4776 │
├──────────┼──────────┼─────────────┼──────────────────┼──────────────────┼──────────────────┼──────────────────┤
│ 4KB      │ Read     │         114 │ 2, 1291, 287     │ 136, 578, 209    │ 0, 5, 0          │ 140, 1553, 497   │
├──────────┼──────────┼─────────────┼──────────────────┼──────────────────┼──────────────────┼──────────────────┤
│ 8KB      │ Read     │           6 │ 413, 646, 505    │ 266, 1043, 482   │ 0, 0, 0          │ 713, 1509, 988   │
├──────────┼──────────┼─────────────┼──────────────────┼──────────────────┼──────────────────┼──────────────────┤
│ 16KB     │ Read     │          18 │ 386, 633, 503    │ 228, 41210, 5068 │ 0, 0, 0          │ 711, 41747, 5572 │
├──────────┼──────────┼─────────────┼──────────────────┼──────────────────┼──────────────────┼──────────────────┤
│ 32KB     │ Read     │          17 │ 31, 748, 478     │ 254, 1455, 557   │ 0, 0, 0          │ 286, 1898, 1036  │
├──────────┼──────────┼─────────────┼──────────────────┼──────────────────┼──────────────────┼──────────────────┤
│ 64KB     │ Read     │          12 │ 377, 626, 498    │ 413, 41218, 4151 │ 0, 0, 0          │ 793, 41765, 4650 │
├──────────┼──────────┼─────────────┼──────────────────┼──────────────────┼──────────────────┼──────────────────┤
│ 128KB    │ Read     │           3 │ 363, 499, 450    │ 654, 1199, 877   │ 0, 0, 0          │ 1154, 1563, 1329 │
╘══════════╧══════════╧═════════════╧══════════════════╧══════════════════╧══════════════════╧══════════════════╛

Total IOs count: 174
```
Now to show it on top tool, I want to calculate stats between two iterations/batches.

But I see that in "connection get_io_statistics", last 4 latency stats columns (bdev, net, qos, total) show all time stats. So to calculate between two iterations, I cannot work with max or min (because max would be all time max latency, and there is no way to get max between an iteration). So I did some calculations with mean values to get mean btw two iterations:

Calculation done for each bucket:
1. total_io_time (t) = mean_latency * io_count 
2. use above total_io_time formula to calculate io time between two iterations (t2 - t1). 
3. mean latency btw two iteration = t2-t1 / (io_count at t2 - io_count at t1)

Then used same calculation for all mean latency (bdev, net, qos, total).

## Output 

First iteration always has garbage values (like other top commands). Also scheduled fio in background on 1  namespace of this subsystems from that host. 
```
[ceph: root@newtop-vallari-cluster-node1 /]# ceph nvmeof top controller 'nqn.2016-06.io.spdk:cnode1.mygroup1' 'nqn.2014-08.org.nvmexpress:uuid:6def1fb3-a3a2-447b-9938-59856fc4805f'
Gateway                  Size   Total IOPS    rIOPS   rBDEV µs   rNet µs   rQoS µs   rTotal µs    wIOPS   wBDEV µs   wNet µs   wQoS µs   wTotal µs
10.243.64.68:5500       128KB          731      731        641       781         0       1424        -          -         -         -          -
10.243.64.69:5500       128KB           46       46        500       958         1       1461        -          -         -         -          -
10.243.64.67:5500       128KB           46       46        453       794         0       1248        -          -         -         -          -
10.243.64.68:5500        16KB       128968    85708        618       272         0        891    43260       2711       205         0       2918
10.243.64.69:5500        16KB          139      139        407       447         0        855        -          -         -         -          -
10.243.64.67:5500        16KB          139      139        596       341         1        940        -          -         -         -          -
10.243.64.68:5500         2KB           58        -          -         -         -          -       58      33011         0         1      33012
10.243.64.69:5500         2KB           46        -          -         -         -          -       46      20732         0         1      20733
10.243.64.67:5500         2KB           35        -          -         -         -          -       35       4314         0         1       4315
10.243.64.68:5500        32KB       192791   127877        666       331         0        999    64914       2980       248         0       3229
10.243.64.69:5500        32KB          186      186        258       450         1        710        -          -         -         -          -
10.243.64.67:5500        32KB          186      186        486       425         0        912        -          -         -         -          -
10.243.64.68:5500         4KB        58528    45513        474       191         0        666    13015       2480         0         1       2481
10.243.64.69:5500         4KB         1869     1869        196       189         1        387        -          -         -         -          -
10.243.64.67:5500         4KB         1637     1637        179       171         0        351        -          -         -         -          -
10.243.64.68:5500        64KB        36236    24777        651       407         0       1060    11459       3619       295         0       3915
10.243.64.69:5500        64KB           93       93        395       549         0        945        -          -         -         -          -
10.243.64.67:5500        64KB           93       93        656       486         0       1144        -          -         -         -          -
10.243.64.68:5500         8KB        73505    49634        579       217         0        797    23871       2595        83         1       2680
10.243.64.69:5500         8KB           46       46        386       441         0        829        -          -         -         -          -
10.243.64.67:5500         8KB           46       46        489       517         0       1007        -          -         -         -          -
 ---- 

Gateway                  Size   Total IOPS    rIOPS   rBDEV µs   rNet µs   rQoS µs   rTotal µs    wIOPS   wBDEV µs   wNet µs   wQoS µs   wTotal µs
10.243.64.68:5500       128KB            0        0          0         0         0          0        -          -         -         -          -
10.243.64.69:5500       128KB            0        0          0         0         0          0        -          -         -         -          -
10.243.64.67:5500       128KB            0        0          0         0         0          0        -          -         -         -          -
10.243.64.68:5500        16KB          135       62        618       272         0        891       72       2854       205         0       3013
10.243.64.69:5500        16KB            0        0          0         0         0          0        -          -         -         -          -
10.243.64.67:5500        16KB            0        0          0         0         0          0        -          -         -         -          -
10.243.64.68:5500         2KB            0        -          -         -         -          -        0          0         0         0          0
10.243.64.69:5500         2KB            0        -          -         -         -          -        0          0         0         0          0
10.243.64.67:5500         2KB            0        -          -         -         -          -        0          0         0         0          0
10.243.64.68:5500        32KB          191       90        666       443         0        999      101       3031       197         0       3280
10.243.64.69:5500        32KB            0        0          0         0         0          0        -          -         -         -          -
10.243.64.67:5500        32KB            0        0          0         0         0          0        -          -         -         -          -
10.243.64.68:5500         4KB           42       18        474       191         0        863       24       2612         0         1       2613
10.243.64.69:5500         4KB            0        0          0         0         0          0        -          -         -         -          -
10.243.64.67:5500         4KB            0        0          0         0         0          0        -          -         -         -          -
10.243.64.68:5500        64KB           35       16        531       407         0        821       19       3619       343         0       3963
10.243.64.69:5500        64KB            0        0          0         0         0          0        -          -         -         -          -
10.243.64.67:5500        64KB            0        0          0         0         0          0        -          -         -         -          -
10.243.64.68:5500         8KB           80       42        485       217         0        703       38       2595        33         1       2680
10.243.64.69:5500         8KB            0        0          0         0         0          0        -          -         -         -          -
10.243.64.67:5500         8KB            0        0          0         0         0          0        -          -         -         -          -
 ---- 

Gateway                  Size   Total IOPS    rIOPS   rBDEV µs   rNet µs   rQoS µs   rTotal µs    wIOPS   wBDEV µs   wNet µs   wQoS µs   wTotal µs
10.243.64.68:5500       128KB            2        2        576       781         0       1326        -          -         -         -          -
10.243.64.69:5500       128KB            0        0          0         0         0          0        -          -         -         -          -
10.243.64.67:5500       128KB            0        0          0         0         0          0        -          -         -         -          -
10.243.64.68:5500        16KB          193      164        702       356         0       1060       29       2714       205         0       3040
10.243.64.69:5500        16KB            0        0          0         0         0          0        -          -         -         -          -
10.243.64.67:5500        16KB            0        0          0         0         0          0        -          -         -         -          -
10.243.64.68:5500         2KB            0        -          -         -         -          -        0          0         0         0          0
10.243.64.69:5500         2KB            0        -          -         -         -          -        0          0         0         0          0
10.243.64.67:5500         2KB            0        -          -         -         -          -        0          0         0         0          0
10.243.64.68:5500        32KB          283      238        753       419         0       1172       45       2981       247         0       3230
10.243.64.69:5500        32KB            0        0          0         0         0          0        -          -         -         -          -
10.243.64.67:5500        32KB            0        0          0         0         0          0        -          -         -         -          -
10.243.64.68:5500         4KB          106      100        511       228         0        704        6       2483         0         1       2649
10.243.64.69:5500         4KB            0        0          0         0         0          0        -          -         -         -          -
10.243.64.67:5500         4KB            0        0          0         0         0          0        -          -         -         -          -
10.243.64.68:5500        64KB           62       50        730       487         0       1219       12       3619       296         0       3995
10.243.64.69:5500        64KB            0        0          0         0         0          0        -          -         -         -          -
10.243.64.67:5500        64KB            0        0          0         0         0          0        -          -         -         -          -
10.243.64.68:5500         8KB          111       91        666       261         0        927       20       2691        82         1       2776
10.243.64.69:5500         8KB            0        0          0         0         0          0        -          -         -         -          -
10.243.64.67:5500         8KB            0        0          0         0         0          0        -          -         -         -          -
  ---- 

^CInterrupted
```

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [x] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins test classic perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-classic/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test crimson perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-crimson/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test signed` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pr-commits/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-commits/config/definitions/ceph-pr-commits.yml)
- `jenkins test make check` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests/config/definitions/ceph-pull-requests.yml)
- `jenkins test make check arm64` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests-arm64/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests-arm64/config/definitions/ceph-pull-requests-arm64.yml)
- `jenkins test submodules` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-pr-submodules/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-submodules/config/definitions/ceph-pr-commits.yml)
- `jenkins test dashboard` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-pull-requests/config/definitions/ceph-dashboard-pull-requests.yml)
- `jenkins test dashboard cephadm` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-cephadm-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml)
- `jenkins test api` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-api/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-api/config/definitions/ceph-pr-api.yml)
- `jenkins test docs` [ReadTheDocs](https://readthedocs.org/projects/ceph/) | [Github Workflow Definition](https://github.com/ceph/ceph/blob/main/.readthedocs.yml)
- `jenkins test ceph-volume all` [Jenkins Jobs](https://jenkins.ceph.com/view/ceph-volume%20PR/) | [Jenkins Jobs Definition](https://github.com/ceph/ceph-build/blob/main/ceph-volume-cephadm-prs/config/definitions/ceph-volume-pr.yml)
- `jenkins test windows` [Jenkins Job](https://jenkins.ceph.com/job/ceph-windows-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-windows-pull-requests/config/definitions/ceph-windows-pull-requests.yml)
- `jenkins test rook e2e` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-orchestrator-rook-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-rook-e2e/config/definitions/ceph-orchestrator-rook-e2e.yml)

You must only issue one Jenkins command per-comment. Jenkins does not understand
comments with more than one command.
</details>

### CODE REVIEW

File: src/pybind/mgr/dashboard/tests/test_nvmeof_top_cli.py [hunk 1]
```diff
@@ -172,6 +181,76 @@ def test_invalid_sort_key(self, io_collector):
             tool.format_output()
 
 
+class TestNVMeoFTopHostControllerFormat:
```
Comment (afreen23) [inline src/pybind/mgr/dashboard/tests/test_nvmeof_top_cli.py]: Format tests look good. Please also add CLI command integration tests for `nvmeof top controller` in `TestNvmeofTopCommands` below -- at minimum `test_top_controller_success`, `test_top_controller_missing_nqn_returns_einval`, `test_top_controller_missing_host_nqn_returns_einval`, and `test_top_controller_run_fail`. See existing `test_top_cpu_*` / `test_top_io_*` tests for the pattern.

**Review by afreen23 — CHANGES_REQUESTED** [review]
Code follows existing patterns well. Missing CLI command integration tests.

File: src/pybind/mgr/dashboard/services/nvmeof_top_cli.py [hunk 2]
```diff
@@ -136,6 +136,36 @@ def calculate(self, delay: float):
...
+            self.bdev_sum = Counter()
+            self.net_sum = Counter()
+            self.qos_sum = Counter()
+            self.total_sum = Counter()
+
+            self.ops_rate = 0.0
+            self.bdev_mean = 0.0
+            self.net_mean = 0.0
+            self.qos_mean = 0.0
+            self.total_mean = 0.0
```
Comment (afreen23) [inline src/pybind/mgr/dashboard/services/nvmeof_top_cli.py:151]: add type annotation

```suggestion
            self.controller_stats: dict[tuple[str, int], dict[str, ControllerBucketStats]] = {}
```

File: src/pybind/mgr/dashboard/tests/test_nvmeof_top_cli.py [hunk 3]
```diff
@@ -6,7 +6,7 @@
 import pytest
 
 from ..services.nvmeof_top_cli import MAX_SESSION_TTL, Counter, \
-    NvmeofTopCollector, NVMeoFTopCPU, NVMeoFTopIO
+    NvmeofTopCollector, NVMeoFTopCPU, NVMeoFTopHostController, NVMeoFTopIO
```
Comment (afreen23) [inline src/pybind/mgr/dashboard/tests/test_nvmeof_top_cli.py]: missing import update

```suggestion
from ..services.nvmeof_top_cli import MAX_SESSION_TTL, Counter, \
    NvmeofTopCollector, NVMeoFTopCPU, NVMeoFTopHostController, NVMeoFTopIO
```

**Review by afreen23 — CHANGES_REQUESTED** [review]

File: src/pybind/mgr/dashboard/tests/test_nvmeof_top_cli.py (hunk 3 above)
Comment (VallariAg) [reply src/pybind/mgr/dashboard/tests/test_nvmeof_top_cli.py]: Sorry, I didn't understand, what is missing here?

File: src/pybind/mgr/dashboard/services/nvmeof_top_cli.py [hunk 4]
```diff
@@ -148,6 +178,7 @@ def __init__(self):
             self.subsystems: Any = None
             self.reactor_stats = {}
             self.iostats = {}
+            self.connection_stats: dict = {}
```
Comment (afreen23) [inline src/pybind/mgr/dashboard/services/nvmeof_top_cli.py]: nit: add full type annotation (carried over from previous review — class was renamed but generics still missing)

```suggestion
            self.connection_stats: dict[tuple[str, int], dict[str, ConnectionBucketStats]] = {}
```

**Review by afreen23 — APPROVED** [review]

File: src/pybind/mgr/dashboard/services/nvmeof_top_cli.py (hunk 4 above)
Comment (VallariAg) [reply src/pybind/mgr/dashboard/services/nvmeof_top_cli.py]: my bad, included the change in new push! thank you!

**Comment by VallariAg** [comment]
Just a rebase to upstream main

**Comment by VallariAg** [comment]
Trivial rebase to upstream main (to include CI fixes)

**Comment by VallariAg** [comment]
jenkins test dashboard cephadm

File: src/pybind/mgr/dashboard/services/nvmeof_top_cli.py [hunk 5]
```diff
@@ -146,8 +176,9 @@ def __init__(self):
             self.namespaces = {}
             self.lbg_to_gateway: dict = {}
             self.subsystems: Any = None
-            self.reactor_stats = {}
-            self.iostats = {}
+            self.reactor_stats: Dict[str, Dict[str, ReactorStats]] = {}
+            self.iostats: Dict[str, Dict[str, PerformanceStats]] = {}
+            self.connection_stats: Dict[Tuple[str, int], Dict[str, ConnectionBucketStats]] = {}
```
Comment (Hezko) [inline src/pybind/mgr/dashboard/services/nvmeof_top_cli.py:181]: what happens if a connection is being disconnected during the polling? or a host was deleted? I think we will keep showing the last recorded value don't we? maybe we should clean this before each run?

File: src/pybind/mgr/dashboard/services/nvmeof_top_cli.py [hunk 6]
```diff
@@ -255,6 +286,27 @@ def get_reactor_data(self, sort_pos: int, reverse_sort: bool):
...
 
+        def get_connection_data(self, sort_pos: int, reverse_sort: bool):
+            conn_data: List[Tuple[Any, ...]] = []
+            for (gw_addr, size_kb), io_stats in self.connection_stats.items():
+                total_iops = 0.0
+                metrics: List[Optional[float]] = []
+                for io_type in ('read', 'write'):
+                    stats = io_stats.get(io_type)
+                    if stats:
+                        stats.calculate(self.delay)
```
Comment (Hezko) [inline src/pybind/mgr/dashboard/services/nvmeof_top_cli.py]: There is a bit of semantic issue here (not sure what to do with it at this point though), we call a "get" function, but the calculate function actually mutating the data of the io_stats, isn't it? one suggestion is to make a copy of the entire dict and just replace them at the end (externally of the call to `get_connection_data`)

File: src/pybind/mgr/dashboard/services/nvmeof_top_cli.py [hunk 7]
```diff
@@ -255,6 +286,27 @@ def get_reactor_data(self, sort_pos: int, reverse_sort: bool):
...
+                        total_iops += stats.ops_rate
+                        metrics += [stats.ops_rate, stats.bdev_mean, stats.net_mean,
+                                    max(0.0, stats.qos_mean), stats.total_mean]
+                    else:
+                        metrics += [None, None, None, None, None]
+
+                conn_data.append((gw_addr, f"{size_kb}KB", total_iops, *metrics))
+
+            conn_data.sort(key=lambda t: t[sort_pos] if t[sort_pos] is not None else 0.0,
```
Comment (Hezko) [inline src/pybind/mgr/dashboard/services/nvmeof_top_cli.py:305]: Im not sure our sort is working well for all columns. as it is lexicographic and not numeric, we might have strange order of the results. for example the size column values: `2KB`, `10KB` and `32KB` would sort like this: "10KB" < "2KB" < "32KB" when sorted lexicographically although numerically it makes no sense.

**Review by Hezko — CHANGES_REQUESTED** [review]
Good job!

**Comment by VallariAg** [comment]
I updated this PR to include fixes for "top connection" command, and I moved fixes generic to other 2 top commands (io/cpu) to https://github.com/ceph/ceph/pull/71468. This is because "top connection" command would only be taken to umbrella/10.0 - so moved generic fixes to PR 71468 so we can backport it to tentacle/9.2 easily too. 
There would be no conflicts between this and PR 71468 - so they can be tested and merge together too.

**Review by Hezko — APPROVED** [review]

**Comment by VallariAg** [comment]
This PR has been added to [wip-vallari-testing-20260903.071327-main](https://tracker.ceph.com/issues/80235).
